### DISCUSSION TITLE
Why we don't have a Operator is use HiveServer2Hook?

### DISCUSSION THREAD

**Discussion by Jie211** [apache/airflow#32483] [Q&A] [answered]
There are two types of Hooks available for Hive.
- HiveServer2Hook: interacting with the HiveServer2 API by using Pyhive library
- HiveCliHook: HiveCli Wrapper
However, it seems that only the HiveOperator use HiveCliHook.

Since HiveServer2Hook inherits from DbApiHook, shouldn't we be using the common-sql Provider's Operator (such as SQLExecuteQueryOperator) when using HiveServer2Hook in an Operator? 

If that is the correct approach, then why doesn't the HiveServer2Hook implement the [run method](https://github.com/apache/airflow/blob/providers-apache-hive/6.1.2/airflow/providers/common/sql/hooks/sql.py#L271-L279)? 

This seems to have an impact on using options like [hive_conf](https://github.com/apache/airflow/blob/providers-apache-hive/6.1.2/airflow/providers/apache/hive/hooks/hive.py#L896) in operators like SQLExecuteQueryOperator. Is this behavior by design or is it an oversight?

**Comment by potiuk** [comment]
> If that is the correct approach, then why doesn't the HiveServer2Hook implement the [run method](https://github.com/apache/airflow/blob/providers-apache-hive/6.1.2/airflow/providers/common/sql/hooks/sql.py#L271-L279)?

Why do you think it should? None of the other DBApi-derived hooks implements run() method AFAIK. But maybe I am missing something?
 
> This seems to have an impact on using options like [hive_conf](https://github.com/apache/airflow/blob/providers-apache-hive/6.1.2/airflow/providers/apache/hive/hooks/hive.py#L896) in operators like SQLExecuteQueryOperator. Is this behavior by design or is it an oversight?

What Impact? What problem do you have? What did you try and did not work?

**Comment by Jie211** [comment]
@potiuk 

Thank you for your attention to this matter.

> Why do you think it should? None of the other DBApi-derived hooks implements run() method AFAIK. But maybe I am missing something?

Yes, I don't think that all DBApi-like hooks need to implement the run method too.

But may be different from other SQL databases, Hive might require hive_conf settings.
Although the HiveServer2Hook has implemented methods to handle hive_conf, if we use common-sql's Operators by using the HiveServer2Hook, we would need to handle hive_conf processing by ourselves.
This is because common-sql's Operators don't use methods like get_results or get_records, they use the (low-level?) run method.

> What Impact? What problem do you have? What did you try and did not work?

I'm not aware of any issues or problems with that. Because if we need some pre-process like hive_conf, i think we can handle this by creating our own Operator or directly using the hook, but what I would like to know is why the HiveServer2Hook was designed this way?

**Reply by potiuk** [reply]
> Although the HiveServer2Hook has implemented methods to handle hive_conf, if we use common-sql's Operators by using the HiveServer2Hook, we would need to handle hive_conf processing by ourselves.

Why can't you use hook_params for that? What's the problem with it?

**Reply by Jie211** [reply]
> Why can't you use hook_params for that? What's the problem with it?

I'm sorry if I misunderstood, 
but I think that hive_conf is not an attribute of the HiveServer2Hook class, so passing hive_conf to hook_params will not make any difference.

Perhaps my explanation was not clear, but when I mention hive_conf, it is not just a parameter i need to pass to hook.
It refers to executing the hive_conf that was passed with `set` , before running the main sql query.
similar to how the other methods implemented in HiveServer2Hook which is accept hive_conf as a parameter.

**Reply by potiuk** [reply, marked as answer]
Feel free to provide a fix for it then. This is likely a bug in implementation.

**Reply by potiuk** [reply]
Just to explain: We have > 2500 contributors and around 100 merged PRs a week. And many thousands of operators. So bugs happen. And trying to find out "why" when you have 1000s operators and  clear mistake, fixing it the best way you can give back for the free software that those > 2500 (mostly volunteering in their free time) contribute so that it can be used for free.

I don't think there is anyone who can authoritatively answer "it is how it is because" -> I think it's a bug and you seem to already know how to fix it. Go for it.

**Reply by Jie211** [reply]
As there are many Hive users, I thought that it might have been my misunderstanding.
Thank you for your confirmation, and I will submit a PR as my first contribution!
